lib/smf/smf.h

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright 2021 The Chromium OS Authors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* State Machine Framework */
+
+#ifndef _SMF_H__
+#define _SMF_H__
+
+#ifndef CONFIG_SMF_FLAT
+/**
+ * @brief Macro to create a hierarchical state that prints a message
+ *        on entry to and exit from the state.
+ *
+ * @param _entry_msg Message printed on state entry
+ * @param _exit_msg  Message printed on state exit
+ * @param _entry     State entry function
+ * @param _run       State run function
+ * @param _exit      State exit function
+ * @param _parent    State parent object or NULL
+ */
+#define HSM_CREATE_MSG_STATE( \
+	_entry_msg, _exit_msg, _entry, _run, _exit, _parent) \
+{ \
+	.entry_msg = _entry_msg, \
+	.exit_msg  = _exit_msg,  \
+	.entry     = _entry,     \
+	.run       = _run,       \
+	.exit      = _exit,      \
+	.parent    = _parent     \
+}
+
+/**
+ * @brief Macro to create a hierarchical state.
+ *
+ * @param _entry  State entry function
+ * @param _run    State run function
+ * @param _exit   State exit function
+ * @param _parent State parent object or NULL
+ */
+#define HSM_CREATE_STATE(_entry, _run, _exit, _parent) \
+{ \
+	.entry  = _entry, \
+	.run    = _run,   \
+	.exit   = _exit,  \
+	.parent = _parent \
+}
+
+#else
```

Comment:
Is it a supported use case to have CONFIG_SMF_FLAT=n, but enable a flat state machine.  In other words, is it supported to enable both hierarchical and flat state machines in the same build?

Reply:
The short answer is no. Initially I was going to just have the hierarchical state and a designer could leave all the parent entries NULL if they wanted a purely flat state machine but I thought that would waste memory, so that's why I added the CONFIG_SMF_FLAT macro.